test/e2e/framework/kubectl/kubectl_builder.go

```diff
@@ -0,0 +1,352 @@
+	b.cmd.Stdin = strings.NewReader(data)
+	return &b
+}
+
+// WithStdinReader sets the given reader and returns itself.
+func (b Builder) WithStdinReader(reader io.Reader) *Builder {
+	b.cmd.Stdin = reader
+	return &b
+}
+
+// ExecOrDie runs the kubectl executable or dies if error occurs.
+func (b Builder) ExecOrDie(namespace string) string {
+	str, err := b.Exec()
+	// In case of i/o timeout error, try talking to the apiserver again after 2s before dying.
+	// Note that we're still dying after retrying so that we can get visibility to triage it further.
+	if isTimeout(err) {
+		framework.Logf("Hit i/o timeout error, talking to the server 2s later to see if it's temporary.")
+		time.Sleep(2 * time.Second)
+		retryStr, retryErr := RunKubectl(namespace, "version")
+		framework.Logf("stdout: %q", retryStr)
+		framework.Logf("err: %v", retryErr)
+	}
+	framework.ExpectNoError(err)
+	return str
+}
+
+func isTimeout(err error) bool {
+	switch err := err.(type) {
+	case *url.Error:
+		if err, ok := err.Err.(net.Error); ok && err.Timeout() {
+			return true
+		}
+	case net.Error:
+		if err.Timeout() {
+			return true
+		}
+	}
+	return false
+}
+
+// Exec runs the kubectl executable.
+func (b Builder) Exec() (string, error) {
+	stdout, _, err := b.ExecWithFullOutput()
+	return stdout, err
+}
+
+// ExecWithFullOutput runs the kubectl executable, and returns the stdout and stderr.
+func (b Builder) ExecWithFullOutput() (string, string, error) {
+	var stdout, stderr bytes.Buffer
+	cmd := b.cmd
+	cmd.Stdout, cmd.Stderr = &stdout, &stderr
+
+	framework.Logf("Running '%s %s'", cmd.Path, strings.Join(cmd.Args[1:], " ")) // skip arg[0] as it is printed separately
+	if err := cmd.Start(); err != nil {
+		return "", "", fmt.Errorf("error starting %v:\nCommand stdout:\n%v\nstderr:\n%v\nerror:\n%v", cmd, cmd.Stdout, cmd.Stderr, err)
+	}
+	errCh := make(chan error, 1)
+	go func() {
+		errCh <- cmd.Wait()
+	}()
+	select {
+	case err := <-errCh:
+		if err != nil {
+			var rc = 127
+			if ee, ok := err.(*exec.ExitError); ok {
+				rc = int(ee.Sys().(syscall.WaitStatus).ExitStatus())
+				framework.Logf("rc: %d", rc)
+			}
+			return stdout.String(), stderr.String(), uexec.CodeExitError{
+				Err:  fmt.Errorf("error running %v:\nCommand stdout:\n%v\nstderr:\n%v\nerror:\n%v", cmd, cmd.Stdout, cmd.Stderr, err),
+				Code: rc,
+			}
+		}
+	case <-b.timeout:
+		b.cmd.Process.Kill()
+		return "", "", fmt.Errorf("timed out waiting for command %v:\nCommand stdout:\n%v\nstderr:\n%v", cmd, cmd.Stdout, cmd.Stderr)
+	}
+	framework.Logf("stderr: %q", stderr.String())
+	framework.Logf("stdout: %q", stdout.String())
+	return stdout.String(), stderr.String(), nil
+}
+
+// RunKubectlOrDie is a convenience wrapper over Builder
+func RunKubectlOrDie(namespace string, args ...string) string {
+	tk := NewTestKubeconfig(framework.TestContext.CertDir, framework.TestContext.Host, framework.TestContext.KubeConfig, framework.TestContext.KubeContext, framework.TestContext.KubectlPath, namespace)
```

Comment:
I think it'd be better to pass this `tk` struct as an argument.
Otherwise we have a dependency in the test framework that is not explicit.

Think it would also be good to refactor how the `NewKubectlCommand` depends on the `tk`s to not have to pass `tk`s are method arguments so much but we could leave that for a follow up PR.